### ISSUE TITLE
Alternate audio devices not remembered correctly

### ISSUE THREAD

**Issue by AtmanActive** [miroslavpejic85/mirotalk#208]
# Bug Report

`IMPORTANT`: We primarily use GitHub as an issue tracker. Just open an issue here if you have encountered a bug in MiroTalk.
If you have `ideas, suggestions, questions, doubts` about MiroTalk or `need support`, please use the MiroTalk Forum instead:

👉 [https://discord.gg/rgGYfeYW3N](https://discord.gg/rgGYfeYW3N)

## Describe the bug

On windows, for users with several audio cards, p2p.mirotalk.com DOES NOT remember the device choice correctly. On first subsequent start, the screen SHOWS correct device selection but the ACTUAL audio is still routed to default audio device (instead of the one previously chosen). On second start, audio devices selection is back to default, as if someone reset the selection behind our backs.

## To reproduce

1. On a windows machine with muliple audio devices, where device A is windows system default and device B is free:
2. Open any chromium based browser, go to p2p.mirotalk.com.
3. Choose audio device B as your audio input and output.
4. Enter the session - everything is working correctly.
5. Now close the tab/browser and start it again.
6. Audio device selection screen looks like it remembered your previous choice, so don't touch anything there, just join the session.
7. No audio! That's because MiroTalk opened audio device A even though it did show that it will open device B.
8. Close the tab/browser and start it again.
9. Audio device selection screen shows device A (default) as selected even though the last time you touched those controls you did set it to use device B.

## Expected behavior

Once I chose an audio device B as my input and output, I expect MiroTalk to automatically use it each and every time I open that same browser.

## Desktop - Mobile

Please complete the following information:

 - Device or OS: Windows
 - Browser: Chrome
 - Version: 120 x64

**Comment by miroslavpejic85** [comment]
Hi @AtmanActive,
Can you please try now? & Thank you!

**Comment by AtmanActive** [comment]
Hi @miroslavpejic85 , 

Thanks for the quick patch.

Unfortunately, now it is even worse. :-)

Now the UI does show properly the remembered choice but the audio engine completely ignores it.

So:
1. We set our choice to audio device B
2. Everything is working as expected.
3. Close browser, start again
4. UI showing audio device B
5. No audio!
6. Close browser, start again
7. UI showing audio device B
8. No audio!
9. The only way to make it work again is to deliberately switch to audio device A, start session (of course, no audio), close the browser, start again, manually set to device B, and thus we are back to point number 2.

At least before, the UI was showing that something is off, after a few restarts, but now the UI is showing that everything is ok, while in fact, there is no audio.

In other words, it works only the first time we set it to audio device B and never again. Until we rejig the UI controls deliberately to make it work again.

Thanks.

**Comment by miroslavpejic85** [comment]
Hi @AtmanActive, you're welcome ;)

From what we've tried here it seems to work, maybe Try to clear the cache to see if it resolves the problem. Here are some steps you can take to troubleshoot and potentially resolve the audio issue:

1. **Clear Browser Cache:**
   - Clear the browser cache to ensure that there are no conflicting or outdated settings causing the problem.

2. **Check Audio Device Settings:**
   - Confirm that the audio device settings in your operating system are correctly configured. Sometimes, issues can arise if there are conflicts or incorrect settings at the system level.

3. **Update Browser and Audio Drivers:**
   - Ensure that your browser is up-to-date. An outdated browser version might have compatibility issues.
   - Similarly, make sure that your audio drivers are up-to-date. Check the manufacturer's website for the latest drivers for your audio device.

4. **Try a Different Browser:**
   - If the issue persists, try using a different browser to see if the problem is specific to the browser you are currently using.

5. **Check for Browser Extensions:**
   - Some browser extensions or add-ons might interfere with audio settings. Disable extensions one by one to identify if any of them are causing the issue.

Let me know and thank to you.

**Comment by miroslavpejic85** [comment]
> Please note that the selected video and audio devices will be retained if the count of it before and after the modification remains the same. If a device is removed or a new one is added, the default video/audio will be preserved.

WIP: I want to enhance the logic by checking if the previous selected ID exists rather than comparing the count of devices. If the previous selected ID does not exist, will be selected to the default one.

**Comment by miroslavpejic85** [comment]
Hi @AtmanActive,

When you have a chance, can you please check if now is better?

Thank you!

**Comment by miroslavpejic85** [comment]
Changing the speakers, the issue persists. O.o
I need to investigate more thoroughly as soon have more time to identify and resolve the problem.

**Comment by miroslavpejic85** [comment]
Hi @AtmanActive, Should be fixed.

**Comment by AtmanActive** [comment]
Sorry, I'm bed ridden with fever at the moment, will test it out as soon as I get back on my feet.

Thanks.

**Comment by miroslavpejic85** [comment]
I'm sorry to hear that you're not feeling well. Take the time you need to rest and recover.

**Comment by miroslavpejic85** [comment]
> \[!NOTE]
>
> In the audio output, you will exclusively hear the voices of other participants. If you find yourself alone in the room, you won't perceive any audio. This design intentionally excludes self-auditory feedback, as it lacks practical significance to hear oneself. :)

**Comment by miroslavpejic85** [comment]
I'm closing the task now as I've tested it and everything seems to be working correctly to not bother. The alternate audio devices are being remembered as expected. I invite anyone in the future to join our [forum](https://discord.gg/rgGYfeYW3N) and discuss there too. 🙏 
Thank you!
